### PR TITLE
Fix bug on Mail.send() when text or input are Unicode

### PR DESCRIPTION
**Pull request by rafaelol**
On PR #964 @matclab forced the encoding of both subject and
text variables to unicode.

After merging it, matclab realized that when we send Unicode
text to the method it raises an exception and asked if he should
change the commit. Unfortunately this thing was kept untouched.

This problem exists because we previously encode the unicode variables
to utf-8 (for instance here https://github.com/web2py/web2py/blob/master/gluon/tools.py#L478-L481) 
and then convert to unicode again. This piece of code shows the exception being raised:

```
>>> a = u'áéí'
>>> a
u'\xe1\xe9\xed'
>>> b = a.encode('utf-8')
>>> b
'\xc3\xa1\xc3\xa9\xc3\xad'
>>> unicode(a)
u'\xe1\xe9\xed'
>>> unicode(b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 0: ordinal not in range(128)
```

If we force to str, just like @matclab suggested, we solve this issue.

### CODE REVIEW

File: gluon/tools.py [hunk 1]
```diff
@@ -777,16 +777,16 @@ def encoded_or_raw(text):
                 if attachments:
                     result = mail.send_mail(
                         sender=sender, to=origTo,
-                        subject=unicode(subject), body=unicode(text), html=html,
+                        subject=str(subject), body=str(text), html=html,
```
Comment (cassiobotaro) [inline gluon/tools.py]: Maybe a better solution is 
```diff
- subject=str(subject), body=str(text), html=html,
+ subject=unicode(subject, encoding), body=unicode(text, encoding), html=html,
```

Comment (rafaelol) [inline gluon/tools.py]: Great! I think it's better than matclab's solution.

Will send another commit to the PR.

Comment (matclab) [inline gluon/tools.py]: Yeap, I've come to the same solution on my local tree, but totally forgot
to submit a new PR…
I'm sorry for that !

Thanks a lot Raphael for reviving it !

-Mathieu

2016-01-07 17:57 GMT+01:00 Rafael Oliveira Lopes <notifications@github.com>:
